gstreamer-sdp/src/s_d_p_message.rs

```diff
@@ -1,11 +1,440 @@
+// Copyright (C) 2018 Sebastian Dröge <sebastian@centricular.com>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use ffi;
+use glib;
+use glib_ffi;
+
+use auto::SDPResult;
+use auto::MikeyMessage;
+use s_d_p_attribute::SDPAttribute;
+use s_d_p_bandwidth::SDPBandwidth;
+use s_d_p_connection::SDPConnection;
+use s_d_p_key::SDPKey;
+use s_d_p_origin::SDPOrigin;
+use s_d_p_time::SDPTime;
+use s_d_p_zone::SDPZone;
+
+glib_wrapper! {
+    pub struct SDPMessage(Boxed<ffi::GstSDPMessage>);
+
+    match fn {
+        copy => |ptr| gobject_ffi::g_boxed_copy(ffi::gst_sdp_message_get_type(), ptr as *mut _) as *mut ffi::GstSDPMessage,
+        free => |ptr| gobject_ffi::g_boxed_free(ffi::gst_sdp_message_get_type(), ptr as *mut _),
+        get_type => || ffi::gst_sdp_message_get_type(),
+    }
+}
 
 impl SDPMessage {
     pub fn new() -> SDPMessage {
         assert_initialized_main_thread!();
         unsafe {
             let msg = glib_ffi::g_malloc(mem::size_of::<ffi::GstSDPMessage>());
-            let _ = ffi::gst_sdp_message_new(msg);
+            ffi::gst_sdp_message_new(msg);
             from_glib_full(msg)
         }
     }
+
+    pub fn init() -> SDPMessage {
+        assert_initialized_main_thread!();
+        unsafe {
+            let msg = glib_ffi::g_malloc(mem::size_of::<ffi::GstSDPMessage>());
+            ffi::gst_sdp_message_new(msg);
+            from_glib_full(msg)
+        }
+    }
+
+    pub fn add_attribute<'a, P: Into<Option<&'a str>>>(&mut self, key: &str, value: P) -> SDPResult {
+        unsafe {
+            from_glib(ffi::gst_sdp_message_add_attribute(self.to_glib_none_mut().0, key.to_glib_none().0, value.into().to_glib_none().0))
+        }
+    }
+
+    pub fn add_bandwidth(&mut self, bwtype: &str, bandwidth: u32) -> SDPResult {
```

Comment:
Should I add a helper function somewhere to do that? Where do you think it should live?

Reply:
It would just mean to do manual implementations of these functions and then doing `if res == ffi::GST_SDP_RESULT_OK { Ok(whatever) } else { Err(()) }`